File: queue/queue.py
# class Queue:
#     def __init__(self, capacity):
#         self.capacity = capacity
#         self.table = [None for _ in range(capacity)]
#         self.rear = 0
#         self.front = 0
#         self.no = 0

#     def is_empty(self):
#         return self.no <= 0
    
#     def is_full(self):
#         return self.no >= self.capacity

#     def enqueue(self, data):
#         if self.is_full():
#             return False
#         self.table[self.rear] = data
#         self.rear += 1
#         self.no += 1
#         if self.rear == self.capacity:
#             self.rear = 0

#     def dequeue(self):
#         if self.is_empty():
#             return False
#         data = self.table[self.front]
#         self.front += 1
#         self.no -= 1
#         if self.front == self.capacity:
#             self.front = 0
#         return data

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedQueue:
    def __init__(self, data):
        self.front = Node(data)
        self.rear = self.front
    
    def is_empty(self):
        return self.front is None

    def enqueue(self, data):
        self.rear.next = Node(data)
        self.rear = self.rear.next
        if self.is_empty():
            self.front = self.rear

    def dequeue(self):
        if self.is_empty():
            return False
        data = self.front.data
        self.front = self.front.next
        return data

File: queue/queue.js
function FixedQueue(capacity) {
    this.capacity = capacity,
    this.que = new Array(this.capacity),
    this.no = 0,
    this.front = 0,
    this.rear = 0,

    function Empty() {
        console.error('큐가 비어 있습니다.');
    },

    function Full() {
        console.error('큐가 가득 차 있습니다.');
    },

    function len() {
        return this.no;
    },

    function isEmpty() {
        return this.no <= 0;
    },

    function isFull() {
        return this.no >= this.capacity;
    },

    function enque(value) {
        if (this.isFull()) {
            throw FixedQueue.Full();
        }

        this.que[this.rear] = value;
        this.rear++;
        this.no++;

        if (this.rear === this.capacity) {
            this.rear = 0;
        }
    },

    function deque() {
        if (this.isEmpty()) {
            throw FixedQueue.Empty();
        }

        const value = this.que[this.front];
        this.front++;
        this.no--;

        if (this.front === this.capacity) {
            this.front = 0;
        }

        return value;
    },

    function peek() {
        if (this.isEmpty()) {
            throw FixedQueue.Empty();
        }

        return this.que[this.front];
    },


    function find(value) {
        for (let i = 0; i < this.no; i++) {
            const idx = (i + this.front) % this.capacity;

            if (this.que[idx] === value) {
                return idx;
            }
        }

        return -1;
    },

    function count(value) {
        let count = 0;

        for (let i = 0; i < this.no; i++) {
            const idx = (i + this.front) % this.capacity;

            if (this.que[idx] === value) {
                count++;
            }
        }

        return count;
    },

    function contains(value) {
        return this.count(value);
    },

    function clear() {
        this.no = 0;
        this.front = 0;
        this.rear = 0;
    },

    function dump() {
        if (this.isEmpty()) {
            throw FixedQueue.Empty();
        } else {
            const printStr = '';
            
            for (let i = 0; i < this.no; i++) {
                const idx = (i + this.front) % this.capacity;

                printStr += `${this.que[idx]} `;
            }

            console.log(printStr);
        }
    }
}
